### ISSUE TITLE
Allow stepping the simulator in paused state

### ISSUE THREAD

**Issue by osrf-migration** [gazebosim/gz-sim#51]
**Original report ([archived issue](https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-gazebo/issues/51)) by Diego Ferigo (Bitbucket: [dgferigo](https://bitbucket.org/%7B994b27a5-7d6d-42b2-8aa7-2ea82fbe7bf6%7D/)).**

----------------------------------------
# Summary

Allow performing single paused blocking simulation steps to update Systems. Systems could for example read new modifications of the ECM and take actions even if the simulated time is not going forward.

# Motivation

In this moment, the `Server` class allows controlling the simulation in a reproducible way \(issuing blocking calls\) only in unpaused state. If the simulator is stepped in paused state, it will loop forever.

I would expect that a single paused step would be executed similarly to unpaused ones.

Example:

```c++
#include <ignition/gazebo/Server.hh>
#include <ignition/gazebo/ServerConfig.hh>


int main(int /*argc*/, char* /*argv*/[])
{
    ignition::gazebo::ServerConfig config;
    config.SetSdfFile("empty.sdf");
    config.SetUseLevels(false);
    config.SetUpdateRate(1000);

    // Create the server
    ignition::gazebo::Server server(config);

    // This is ok
    std::cout << "Running an unpaused blocking iteration..." << std::endl;
    server.Run(/*blocking=*/true, /*iterations=*/1, /*paused=*/false);
    std::cout << "... done" << std::endl;

    // This is looping forever
    std::cout << "Running a paused blocking iteration..." << std::endl;
    server.Run(/*blocking=*/true, /*iterations=*/1, /*paused=*/true);
    std::cout << "... done" << std::endl;

    return 0;
}
```

```
cmake_minimum_required(VERSION 3.12)
project(Test VERSION 0.1)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

find_package(ignition-gazebo3 COMPONENTS all REQUIRED)

add_executable(Test test.cpp)
target_link_libraries(Test PUBLIC ignition-gazebo3::core)
```

# Describe alternatives you've considered

None

# Additional context

This is somehow related to [https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-gazebo/pull-requests/556](https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-gazebo/pull-requests/556). Executing paused steps could be now useful to process ECM updates without executing a physics step.

**Comment by diegoferigo** [comment]
As discussed in https://github.com/ignitionrobotics/ign-gazebo/pull/61#issuecomment-655621591, the simple workaround of returning after executing blocking paused steps through the existing API is not compatible with its expected semantic that is iterating the simulation. A paused step, contrarily, acts only on the ECS without triggering the physics engine.

https://github.com/ignitionrobotics/ign-gazebo/blob/72e931494d8dc0fa11b81afa6050c7486c55f385/include/ignition/gazebo/Server.hh#L134-L136

In order to allow this use case, that could be useful to update the ECS to process new entities and components without the need to step the physics, a new method should be introduced in the `Server` class. @chapulina @azeey do you have already anything in mind?

Few questions. If we clone 1:1 the simulation iterations with something like (only used for brainstorming, I don't think this is the right signature nor name xD):

```cpp
// _paused=true is implied
RunECS(const bool _blocking = false, const uint64_t _iterations = 0)
```

- do you see any use-case of non-blocking ECS steps?
- do you see any use case of executing multiple paused ECS iterations?

At this stage, I'm not seeing any. It could happen that a particular system requires more than one paused step to get initialized, why not, but in that use case the user could call multiple times the method. The `_iterations` field seems more useful to me for simulation iterations, where the physics is typically stepped at a higher rate (e.g. 1000Hz) than the user code loop (e.g. 25Hz).

**Comment by chapulina** [comment]
> do you see any use-case of non-blocking ECS steps?

I can't think of one right now, I think we could leave this use case out until there's a concrete use.

> do you see any use case of executing multiple paused ECS iterations?

I agree with you that there may be systems that need multiple steps for initialization, and I agree with your conclusion that there are other ways of handling this.

---

How about something like this (after writing, I realized I chose the only argument you hadn't chosen :sweat_smile: ):

```.cpp
      /// \brief Run the server once, all systems will be updated once and then 
      /// this returns. This is a blocking call.
      /// \param[in] _paused True to run simulation in a paused state, false to
      /// run simulation unpaused. If unpaused, the simulation iterations will
      /// be increased by 1.
      /// \return False if the server was terminated before completing
      /// not be initialized, and if the server is already running.
      public: bool RunOnce(const bool _paused = true);
```

We haven't been exposing the "ECS" in the API, so I think that we could continue like this a bit longer. Also refer to [this](https://osrf-migration.github.io/ignition-gh-pages/#!/ignitionrobotics/ign-gazebo/pull-requests/33/page/1#comment-74945039) previous conversation. I hope the semantics of `RunOnce` is clear enough? I'm not 100% convinced.

**Comment by diegoferigo** [comment]
> How about something like this (after writing, I realized I chose the only argument you hadn't chosen :sweat_smile: ):

My questions were aimed at removing both the arguments of https://github.com/ignitionrobotics/ign-gazebo/issues/51#issuecomment-655979495, I was not convinced in the first place to keep them ;) 

I like the idea of `RunOnce`. I gave a thought about the duplication of the logic for running the simulation with two different methods (`Run` and the proposed `RunOnce`), and despite at first I asked myself if it's really necessary and I was wondering if it would not create some confusion, I ended up thinking that it is instead a good idea. I have the feeling that many users, especially who directly runs the simulation operating from C++ and using the `Server` class, might probably use this new `RunOnce` even more than the original `Run`. In fact, issuing multiple iterations per run is not necessary in most cases (and it's also quite uncommon to find in other simulators, even though I think is very useful). In this case the duplication has its own reason to exist.

Something that could be not clear enough, and can be solved by extending the documentation, is the real difference between calling `RunOnce(/*_paused=*/true)` and `Run(/*_blocking=*/true, /*_iterations=*/1, /*_paused=*/true)`. If this get clarified, I think we're on the right way :)

**Comment by chapulina** [comment]
@azeey , since you were also reviewing #61 , I'd like to get your thoughts on this too. Thanks!

**Comment by azeey** [comment]
Without exposing the existence of the ECS, it is difficult to be clear about what this function would do. `RunOnce` seems like a good name, but it's not clear what the `Once` is. If we really must hide the ECS, I can't think of a better name. In that case, I guess we would have
  * `RunOnce(paused=false)` = `Run(blocking=true, iterations=1, paused=false)` 
  * `RunOnce(paused=true)`: like `Run`, but with 0 iterations. This is different from `Run(blocking=true, iterations=0, paused=true)` because passing `iterations=0` is treated as infinite iterations.

**Comment by diegoferigo** [comment]
> Without exposing the existence of the ECS, it is difficult to be clear about what this function would do. If we really must hide the ECS, I can't think of a better name.

I opened #248 on this regard. I remember few conversations in the past about exposing the ECS but it was never ticketed.

> because passing iterations=0 is treated as infinite iterations

Thanks for remembering this! I remember there was a key difference when I was investigating this use-case but then I forgot of this specific case. Since we're still brainstorming, do you have already something in mind in the case the existence of the ECS would be exposed? Just a different naming?

**Comment by azeey** [comment]
I believe this issue is resolved by #297.
